source/common/router/config_impl.cc

```diff
@@ -534,6 +533,10 @@ RouteEntryImplBase::RouteEntryImplBase(const VirtualHostImpl& vhost,
                           validator),
       route_name_(route.name()), time_source_(factory_context.mainThreadDispatcher().timeSource()),
       random_value_header_name_(route.route().weighted_clusters().header_name()) {
+  if (route.has_metadata()) {
+    metadata_ = std::make_unique<envoy::config::core::v3::Metadata>(route.metadata());
+    typed_metadata_ = std::make_unique<RouteTypedMetadata>(route.metadata());
+  }
```

Comment:
I find it interesting that both wrap the same metadata field.
I guess this is intentional (after reading this [comment thread](https://github.com/envoyproxy/envoy/pull/4723/files#r225213097)), but just wondering whether there should be some ordering between these, or at least both should not store the same metadata.

Reply:
Yes, I find it strange as well, but it does keep the same metadata in two forms -- as a raw protobuf, and converted to "typed" data structure. I am not aiming to change this here, just avoid creation of both when there is no actual metadata.
To make things even more confusing, RDS keeps a full copy of the entire raw protobuf anyway, so a raw copy of the metadata message here is pure overhead. Again, not for this PR.